text/0000-unsafe-enums.md

```diff
@@ -0,0 +1,99 @@
+- Start Date: 2014-01-23
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add an unsafe enum which is an enum without a discriminant.
+
+# Motivation
+
+When working with FFI, many C libraries will take advantage of unions. Unfortunately Rust has no way
+to represent this sanely, so developers are forced to write duplicated struct definitions for each
+union variant and do a lot of ugly transmuting. Unsafe enums are effectively unions, and allow FFI
+that uses unions to be significantly easier. The syntax chosen here replicates existing syntax without adding new keywords and is thus backwards compatible with existing code.
+
+# Detailed design
+
+An unsafe enum is equivalent to a safe enum except that it does not have a discriminant.
+
+## Declaring an unsafe enum
+
+```rust
+unsafe enum MyEnum {
+    Variant1(c_int),
+    Variant2(*mut c_char),
+    Variant3 {
+        x: f32,
+        y: f32,
+    },
+}
+```
+
+## Instantiating unsafe enums
+This is completely safe.
+```rust
+let foo = Variant1(5);
+```
+
+## Destructuring
+
+Due to the lack of a discriminant, destructuring becomes irrefutable, but also unsafe. Therefore all destructuring needs to be in unsafe functions or blocks. Because destructuring is irrefutable you can directly destructure unsafe enums which isn't possible with safe enums:
+```rust
+unsafe {
+    let Variant1(bar) = foo;
+}
+```
```

Comment:
Perhaps something like `let unsafe { Variant1(bar) } = foo;`? I'm not sure how feasible that would be to implement though.
If we went with the super restrictive trait as suggested by @eddyb then we'd be able to remove the `unsafe` blocks entirely making the resulting code much more concise.